dso/model.py

```diff
@@ -39,37 +39,46 @@ def has_scan_info(datasource: str) -> bool:
 
 def normalise_artefact_extra_id(
     artefact_extra_id: dict[str, str],
-    artefact_version: str=None,
 ) -> str:
     '''
-    generate stable representation of `artefact_extra_id` and remove `version` key if
-    the specified version is identical to the given artefact version
+    generate stable representation of `artefact_extra_id`
 
     sorted by key in alphabetical order and concatinated following pattern:
     key1:value1_key2:value2_ ...
     '''
-    if (version := artefact_extra_id.get('version')) and version == artefact_version:
-        artefact_extra_id = artefact_extra_id.copy()
-        del artefact_extra_id['version']
-
     s = sorted(artefact_extra_id.items(), key=lambda items: items[0])
     return '_'.join([':'.join(values) for values in s])
 
 
-@dataclasses.dataclass(frozen=True)
+@dataclasses.dataclass
 class LocalArtefactId:
-    artefact_name: str | None
-    artefact_type: str | None
+    artefact_name: str | None = None
+    artefact_type: str | None = None
     artefact_version: str | None = None
     artefact_extra_id: dict = dataclasses.field(default_factory=dict)
 
-    def normalised_artefact_extra_id(
-        self,
-        remove_duplicate_version: bool=False,
-    ) -> str:
-        return normalise_artefact_extra_id(
-            artefact_extra_id=self.artefact_extra_id,
-            artefact_version=self.artefact_version if remove_duplicate_version else None,
+    @property
+    def normalised_artefact_extra_id(self) -> str:
+        return normalise_artefact_extra_id(self.artefact_extra_id)
+
+    def as_frozenset(self) -> frozenset[str]:
+        return frozenset((
+            self.artefact_name,
+            self.artefact_type,
+            self.artefact_version,
+            self.normalised_artefact_extra_id,
+        ))
+
+    def __hash__(self) -> int:
+        return hash(self.as_frozenset())
+
+    def __eq__(self, other: 'LocalArtefactId') -> bool:
+        return self.as_frozenset() == other.as_frozenset()
```

Comment:
I think you need to guard against `other` having different type (which might not have `as_frozenset`)